File: C/huffman/huffman_tree.c
#include <stdio.h>
#include "huffman_tree.h"
#include "huffman_node.h"
#include "nodes_list.h"
#include <string.h>

#define SUCCESS 0
#define MEMORY_ERROR -1;

const char CHARS[100] = {
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
        'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
        'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '!', '"', '#', '$', '%', '&', '\'', '(', ')', '*',
        '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`',
        '{', '|', '}', '~', ' ', '\t', '\n', '\r', '\x0b', '\x0c',
};

const int FREQUENCIES[100] = {
        325, 982, 373, 343, 107, 94, 73, 45, 47, 966, 245509, 46768, 67194, 134216, 406157, 69103,
        57328, 218875, 199130, 2788, 29345, 146532, 95890, 216805, 282560, 46849, 2414, 209907,
        215605, 291243, 115225, 34077, 73155, 4681, 85549, 1098, 44560, 15450, 21526, 15726, 42703,
        11770, 11261, 18486, 55893, 2071, 6206, 23919, 15887, 27393, 33268, 12038, 1179, 29040, 34082,
        39878, 14169, 3587, 16508, 608, 9128, 532, 8840, 488, 3, 2, 1, 21, 31077, 643, 644, 73, 0,
        83315, 8150, 78216, 48, 1843, 17199, 468, 0, 440, 10476, 3, 2076, 0, 2068, 0, 71, 1, 0, 33,
        2, 1, 1285884, 0, 124787, 0, 0, 0,
};


huffman_nodes_list_t *create_leaf_nodes() {
    huffman_nodes_list_t *list = huffman_nodes_list_create();
    if (list == NULL) {
        return NULL;
    }

    for (int i = 0; i < 100; i++) {
        huffman_node_t *leaf_node = huffman_node_create(CHARS[i], FREQUENCIES[i]);
        if (leaf_node == NULL) {
            huffman_nodes_list_destroy(list);
            return NULL;
        }

        huffman_nodes_list_add_node(list, leaf_node);
    }
    return list;
}

int huffman_tree_init(huffman_tree_t *tree) {
    huffman_nodes_list_t *list = create_leaf_nodes();
    if (list == NULL) {
        return MEMORY_ERROR;
    }
    while (huffman_nodes_list_length(list) > 1) {
        huffman_node_t *node1 = huffman_nodes_list_get_min_node(list);
        huffman_node_t *node2 = huffman_nodes_list_get_min_node(list);
        huffman_node_t *merged_node = huffman_node_merge_nodes(node1, node2);

        if (merged_node == NULL) {
            huffman_nodes_list_destroy(list);
            return MEMORY_ERROR;
        }

        huffman_nodes_list_add_node(list, merged_node);
    }

    huffman_node_t *root = huffman_nodes_list_get_min_node(list); // last node
    huffman_nodes_list_destroy(list);

    tree->root = root;
    return SUCCESS;
}

char huffman_tree_get_letter(const huffman_tree_t* self, const char* tree_code) {
    return huffman_node_get_letter(self->root, tree_code);
}

bytes_vector_t* huffman_tree_get_code(const struct huffman_tree* self, char letter) {
    return huffman_node_get_code(self->root, letter);
}

void huffman_tree_release(huffman_tree_t *tree) {
    huffman_node_destroy(tree->root);
}

void huffman_tree_print(huffman_tree_t *tree) {
    printf("_____________\n");
    printf("Huffman Tree\n");
    printf("_____________\n");
    huffman_node_print(tree->root);
    printf("_____________\n");
}

File: Rust/src/huffman/huffman_tree.rs
#![allow(warnings, unused)]

use crate::huffman::huffman_node::HuffmanNode;
use crate::huffman::nodes_list::NodesList;
use crate::huffman::InexistentLetterError;

const SIZE: usize = 100;

const CHARS: [char; SIZE] = [
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
    'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
    'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '!', '"', '#', '$', '%', '&', '\'', '(', ')', '*',
    '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`',
    '{', '|', '}', '~', ' ', '\t', '\n', '\r', '\x0b', '\x0c',
];

const FREQUENCIES: [i32; SIZE] = [
    325, 982, 373, 343, 107, 94, 73, 45, 47, 966, 245509, 46768, 67194, 134216, 406157, 69103,
    57328, 218875, 199130, 2788, 29345, 146532, 95890, 216805, 282560, 46849, 2414, 209907,
    215605, 291243, 115225, 34077, 73155, 4681, 85549, 1098, 44560, 15450, 21526, 15726, 42703,
    11770, 11261, 18486, 55893, 2071, 6206, 23919, 15887, 27393, 33268, 12038, 1179, 29040, 34082,
    39878, 14169, 3587, 16508, 608, 9128, 532, 8840, 488, 3, 2, 1, 21, 31077, 643, 644, 73, 0,
    83315, 8150, 78216, 48, 1843, 17199, 468, 0, 440, 10476, 3, 2076, 0, 2068, 0, 71, 1, 0, 33,
    2, 1, 1285884, 0, 124787, 0, 0, 0,
];

pub struct HuffmanTree {
    root: Box<HuffmanNode>
}

impl HuffmanTree {
    pub fn new() -> HuffmanTree {
        let mut nodes_list = HuffmanTree::create_leaf_nodes();
        while nodes_list.length() > 1 {
            let node1 = nodes_list.get_min_node();
            let node2 = nodes_list.get_min_node();
            let merged_node = HuffmanNode::merge_nodes(node1, node2);
            nodes_list.add_node(merged_node);
        }

        let root = nodes_list.get_min_node(); // last node
        return HuffmanTree {
            root
        };
    }

    fn create_leaf_nodes() -> NodesList {
        let mut nodes_list = NodesList::new();
        for i in 0..SIZE {
            let leaf_node = HuffmanNode::new_leaf(CHARS[i], FREQUENCIES[i]);
            nodes_list.add_node(leaf_node);
        }
        return nodes_list;
    }

    /*
     * Given a code, it returns it's associated letter. If the code corresponds to a non-leaf node
     * (therefore it does not contain a letter) then it returns None.
     */
    pub fn get_letter(&self, tree_code: &String) -> Option<char> {
        self.root.get_letter(&mut tree_code.chars())
    }

    /*
     * Given a letter, returns the corresponded code. If the letter does not exist it returns
     * an InexistentLetterError.
     */
    pub fn get_code(&self, letter: char) -> Result<String, InexistentLetterError> {
        return match self.root.get_code(letter) {
            //Ok(code) => code.graphemes(true).rev().collect(),
            Ok(code) => Ok(code.chars().rev().collect::<String>()),
            Err(_) => Err(InexistentLetterError),
        }
    }

    pub fn print(&self) {
        println!("_____________");
        println!("Huffman Tree");
        println!("_____________");
        self.root.print();
        println!("_____________");
    }
}